### PR TITLE
Ports tg's associative-list research datums

### PR DESCRIPTION
**Pull request by Crazylemon64**
Ports tgstation/tgstation#15394

As a result, operations that formerly looped through all research now are constant-time operations.

Also slightly refactors research datums to simplify transferring research data

### CODE REVIEW

**Comment by PsiOmegaDelta** [comment]
Well, assuming it's based on BYOND associative operations it's O(log n) time rather than O(1) but still beats O(n).
Associative lists are red-black trees rather than something that uses hash lookups.

File: code/modules/research/research.dm [hunk 1]
```diff
@@ -116,25 +117,31 @@ research holder datum.
 	for(var/datum/design/PD in possible_designs)
 		if(DesignHasReqs(PD))
 			AddDesign2Known(PD)
-	for(var/datum/tech/T in known_tech)
+	for(var/v in known_tech)
+		var/datum/tech/T = known_tech[v]
+		// On tg, this is 1-20, not sure what that is about
```
Comment (Fox-McCloud) [inline code/modules/research/research.dm]: Only in that PR; a future PR changed it to what we have---it added in abductor tech: https://github.com/tgstation/tgstation/pull/16006

Comment isn't necessary.

Comment (Iamgoofball) [inline code/modules/research/research.dm]: sometimes I whip my weiner around in circles hoping to get it going fast enough to fly away like a helicopter

**Comment by Fox-McCloud** [comment]
Conflicts.

**Comment by Crazylemon64** [comment]
@Fox-McCloud No more conflicts.

**Pull request merged**